### ISSUE TITLE
lxc.sh fails

### ISSUE THREAD

**Issue by jhalterman** [abailly/jepsen-vagrant#3]
I may have filed this and issue #2 together, but when vagrant up fails for various reasons, some of those reasons involve lxc.sh, where the containers are created with some errors and fail to ever successfully start. Depending on the workarounds I try, things continue to fail for different reasons, but initially it's something like (when running in foreground mode):

```
lxc-start: cgroupfs failed to detect cgroup metadata
lxc-start: failed initializing cgroup support
```

**Comment by abailly** [comment]
Hello
Thanks a lot for your interest. It's been a while since I have worked on this project but I was planning to improve it because I will need Jepsen's testing capabilities in my day job. I will have a look at this issue next week.

**Comment by jhalterman** [comment]
Great - thanks!

**Comment by justizin** [comment]
I've had some success the past couple of days on a fresh setup at my new day job, but have you tried logging into the vm with 'vagrant ssh', cd to /vagrant, and manually running lxc.sh?

Also, it looks like jepsen is maybe going under a reorg, so you have to run lein from /jepsen/jepsen.  That is true as of like, an hour ago, could change in days.

**Comment by jhalterman** [comment]
I have tried running manually. The problem I'm running into the lxc setup just fails for various reasons.

**Comment by justizin** [comment]
that's really odd.  have you considered blowing away your vm with 'vagrant destroy' and starting anew?

Again, I have done this successfully on a computer I just started using four days ago.  It is curious that you're getting errors related to things like cgroups that would normally make me think you're trying to do something incompatible with your linux environment, but it's the bundled VM.

I will say, take a look at the pull req I just submitted, #4, though I've had success with the custom vagrant box referenced here, I already have other vagrant envs using debian jessie, so I changed to use that upstream box.

My thinking was, "The change didn't break anything", but maybe it improved things, since the custom box referenced here is getting old and debian/jessie64 is likely updated periodically.

**Comment by jhalterman** [comment]
I tried starting from scratch a few times with no luck. I started to wonder the same thing as you. I'm running on OSX, where vagrant uses virtualbox. I'm not sure to what extent that might effect how things work within the VM. I'll give it one more shot though from scratch and capture the error messages as they occur...

**Comment by jhalterman** [comment]
Ok, here's a rundown of my vagrant up experience...

After running `vagrant up` the shell eventually turns to a bunch of gibberish so it's hard to tell which commands completed successfully. But judging from what happens next, I'm guessing the initial `vagrant up` does not go well. After running `reset` I SSH into the box and try:

```
vagrant@jepsen:/vagrant$ java -v
-bash: java: command not found
```

So setup doesn't appear to have succeeded. Let's try it again:

```
vagrant@jepsen:/vagrant$ sudo bash setup.sh
...
You might want to run 'apt-get -f install' to correct these:
The following packages have unmet dependencies:
 libc-bin : Depends: libc6 (< 2.19) but 2.19-18 is to be installed
 libc-dev-bin : Depends: libc6 (< 2.19) but 2.19-18 is to be installed
 libc6-dev : Depends: libc6 (= 2.18-5) but 2.19-18 is to be installed
 locales : Depends: libc-bin (> 2.19) but 2.18-5 is to be installed
 oracle-java8-installer : Depends: java-common (>= 0.24) but it is not going to be installed
                          Recommends: gsfonts-x11 but it is not going to be installed
E: Unmet dependencies. Try 'apt-get -f install' with no packages (or specify a solution).
...
```

No good. Let's try to fix it:

```
vagrant@jepsen:/vagrant$ sudo apt-get -f install
```

Looks good. Let's try this again:

```
vagrant@jepsen:/vagrant$ sudo bash setup.sh
```

Looking better... i can see JDK8 actually install.

Now let's try this:

```
vagrant@jepsen:/vagrant$ sudo bash net.sh
error: Failed to destroy network default
error: Requested operation is not valid: network 'default' is not active

Network default has been undefined

Network default defined from /tmp/default.xml

starting virtual network default
Network default started
```

Hmm, doesn't seem entirely good? Let's move on to see how it goes:

```
vagrant@jepsen:/vagrant$ sudo bash lxc.sh
mount: cgroup is already mounted or /sys/fs/cgroup busy
       cgroup is already mounted on /sys/fs/cgroup
Kernel configuration not found at /proc/config.gz; searching...
Kernel configuration found at /boot/config-3.13-1-amd64
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Network namespace: enabled
Multiple /dev/pts instances: enabled

--- Control groups ---
Cgroup: enabled
Cgroup namespace: required
```

That last line doesn't seem right. More output:

```
waiting for n1 to be up..........................
```

It never comes up. Let's try to manually start it in foreground mode:

```
vagrant@jepsen:/vagrant$ sudo lxc-start -n n1
lxc-start: cgroupfs failed to detect cgroup metadata
lxc-start: failed initializing cgroup support
lxc-start: failed to spawn 'n1'
lxc-start: The container failed to start.
lxc-start: Additional information can be obtained by setting the --logfile and --logpriority options.
```

No good...

**Comment by abailly** [comment]
Hi @jhalterman I am sorry to hear you are having such difficulties running this project. I am currently testing @bitmonk 's patch to debian/jessie on my Mac. Will report my findings in the weekend.

**Comment by abailly** [comment]
As mentioned in #4 I managed to get `vagrant up` working without any other manual intervention. I am running vagrant 1.6.3 on Mac OS X 10.9.5.

**Comment by jhalterman** [comment]
I just upgraded vagrant (along with virtualbox) and the result is much better!! The containers are up, save for a few errors:

```
==> jepsen: ssh-keygen: /home/vagrant/.ssh/known_hosts: No such file or directory
...
==> jepsen: link /home/vagrant/.ssh/known_hosts to /home/vagrant/.ssh/known_hosts.old: Operation not permitted
```

I'd suggest adding a note to the README about using vagrant >= 1.6.3.

**Comment by jhalterman** [comment]
Running lxc.sh again, the containers get created better, but running jepsen under the vagrant user, things fail with `com.jcraft.jsch.JSchException: Auth fail`.

**Comment by justizin** [comment]
The Jepsen README tells you what to do about that, but I think the jepsen-vagrant also does.

This basically usually means that either the ssh key for vagrant is setup incorrectly (less likely) or that the main host hasn't pulled the host keys for the LXC VMs.

**Comment by jhalterman** [comment]
@bitmonk I made sure that the container ssh keys are in my known_hosts:

```
for i in 1 2 3 4 5; do ssh-keyscan -t rsa n${i}; done >> ~/.ssh/known_hosts
```

I also checked that the vagrant ssh key made it to the containers:

```
vagrant@jepsen:/jepsen$ sudo cat /var/lib/lxc/n1/rootfs/root/.ssh/authorized_keys
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDiwFm+8lypvysSqVfWArAGEmcgNqAbizUq6SWlio0K6Cn5ilZ8sb87DfRW9Ag99atVSjiZZBX386MU2a4K9OL/Fba4E5ddAdLw2rb5mDBxAeRs1mIxnZyfHdEs1Kqt4M08zUVJTnchcVApJMhIjNjVshgQ5Kg/n+ZfVFNWqB/qemTRtJpgiEFX59mSg3zZZuG7uT+X9Fz2ko5t2fi5o/eKulIqovPzpBmyXhY/jexlKXq6GpvokLXtscMGvlKZ7y/2mvMa6KPRkvNPrU7eORsCHcV1IpVayC/An1eRMRLtAz8QJ70fdRFXO6eL3mcBJTz2zCwzqdBTYrgKvs5LJ/jz vagrant@jepsen
```

I noticed that I can SSH to the containers using `ssh root@n1`, but if I try to SSH any other way that requires entering a credential (`root/root`, which is what jepsen does), then things fail:

```
vagrant@jepsen:~$ sudo ssh n1
root@n1's password: 
Permission denied, please try again.
```

Even if I manually set the password, as the jepsen docs suggest:

```
vagrant@jepsen:~$ ssh root@n1
...
root@n1:~# passwd
Enter new UNIX password: root
Retype new UNIX password: root
passwd: password updated successfully
root@n1:~# exit
logout
Connection to n1 closed.
vagrant@jepsen:~$ sudo ssh n1
root@n1's password: root 
Permission denied, please try again.
```

Any ideas?

**Comment by abailly** [comment]
Hello Jonathan,
I modified the authentication parameters in order to use a key for login. I think you can do it at the level of each tests but more simply you can try to set the following in `src/jepsen/control.clj`:

```
(def ^:dynamic *private-key-path*         "SSH identity file"     "/home/vagrant/.ssh/id_rsa")
```

**Comment by jhalterman** [comment]
Thanks - doing that for now. Would be great if we could get the root/root credential'ed login to work. I tried various `sshd_config` configurations, but haven't gotten it to work yet.

**Comment by abailly** [comment]
Did you have some debug trace on the server(s)? That might give us clues on why it fails to properly use root passwords. ssh has a proper verbose that's very helpful and there is something like that in jsch but it needs to be programmed explicitly I think, so requires modification to jepsen's source code.

**Comment by justizin** [comment]
@jhalterman have you given up, here?  I was just catching up on the comments and am really surprised you are continuing to have issues on OSX, which is one of the platforms I haven't had any trouble on.

Also noticed:

```
I'd suggest adding a note to the README about using vagrant >= 1.6.3.
```

I agree with that, have you considered throwing a PR?

I'm working on a chef-based provisioner for a jepsen env, fyi, which may not be everyone's cup of tea, but is actually a LOT less original code, and I think will help pave the way to use jepsen on platforms other than Debian 8, simply because 90-some-% of the chef code will 'just work' elsewhere.

**Comment by jhalterman** [comment]
I eventually got my vagrant environment working great, though if I recall I could never get root/root logins to work. As @abailly suggested, I tried enabling SSH tracing, but didn't find anything.

Lately though, I've been using the [Dockerized Jepsen](https://github.com/aphyr/jepsen/tree/master/docker) setup, which works good for me as well.

Seeing as how this issue is still open and the original cause has been identified (use newer Vagrant), I'll close it for now!
